File: src/main/java/com/anod/appwatcher/model/AddWatchAppAsyncTask.java
package com.anod.appwatcher.model;

import android.content.Context;
import android.os.AsyncTask;
import android.support.v4.util.SimpleArrayMap;

import com.google.android.finsky.api.model.Document;

public class AddWatchAppAsyncTask extends AsyncTask<Document, Void, SimpleArrayMap<String, Integer>> {
    private final WatchAppList mNewAppHandler;
    private final Listener mListener;
    private Context mContext;

    public interface Listener {
        void onAddAppTaskFinish(SimpleArrayMap<String, Integer> result);
    }

    public AddWatchAppAsyncTask(WatchAppList newAppHandler, Context context, Listener listener) {
        mNewAppHandler = newAppHandler;
        mListener = listener;
        mContext = context;
    }

    @Override
    protected SimpleArrayMap<String, Integer> doInBackground(Document... documents) {
        mNewAppHandler.attach(mContext);
        SimpleArrayMap<String, Integer> result = new SimpleArrayMap<>();
        for (Document doc : documents) {
            if (isCancelled()) {
                return result;
            }
            final AppInfo info = new AppInfo(doc);
            int status = mNewAppHandler.addSync(info);
            result.put(info.packageName, status);
        }
        mNewAppHandler.detach();
        return result;
    }

    @Override
    protected void onPostExecute(SimpleArrayMap<String, Integer> result) {
        mListener.onAddAppTaskFinish(result);
    }
}

File: src/main/java/com/anod/appwatcher/model/AddWatchAppAsyncTask.kt
package com.anod.appwatcher.model

import android.annotation.SuppressLint
import android.content.Context
import android.os.AsyncTask
import android.support.v4.util.SimpleArrayMap

import finsky.api.model.Document

class AddWatchAppAsyncTask(
        context: Context,
        private val appList: WatchAppList,
        private val listener: AddWatchAppAsyncTask.Listener) : AsyncTask<Document, Void, SimpleArrayMap<String, Int>>() {

    @SuppressLint("StaticFieldLeak")
    private val context = context.applicationContext

    interface Listener {
        fun onAddAppTaskFinish(result: SimpleArrayMap<String, Int>)
    }

    override fun doInBackground(vararg documents: Document): SimpleArrayMap<String, Int> {
        appList.attach(context)
        val result = SimpleArrayMap<String, Int>()
        for (doc in documents) {
            if (isCancelled) {
                return result
            }
            val info = AppInfo(doc)
            val status = appList.addSync(info)
            result.put(info.packageName, status)
        }
        appList.detach()
        return result
    }

    override fun onPostExecute(result: SimpleArrayMap<String, Int>) {
        listener.onAddAppTaskFinish(result)
    }
}